### PR TITLE
Add initial Gitlab support

### PR DESCRIPTION
**Pull request by lievendoclo**

### CODE REVIEW

**Comment by ddgenome** [comment]
The name of this PR doesn't seem quite right.

**Comment by cdupuis** [comment]
We will merge that after the 1.0 release.

File: lib/operations/common/AbstractRemoteRepoRef.ts [hunk 1]
```diff
@@ -58,7 +59,7 @@ export abstract class AbstractRemoteRepoRef implements RemoteRepoRef {
      * @param {string} sha
      * @param {string} path
      */
-    protected constructor(public readonly providerType: ProviderType,
+    protected constructor(public readonly providerType: ProviderType | string,
```
Comment (cdupuis) [inline lib/operations/common/AbstractRemoteRepoRef.ts]: Why is that a string? GitLab should be added to `ProviderType`.

File: lib/operations/common/GitlabRepoRef.ts [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+            owner,
+            repo,
+            sha,
+            path);
+    }
+
+    public async createRemote(creds: ProjectOperationCredentials, description: string, visibility): Promise<ActionResult<this>> {
+
+        const gitlabUrl = this.concatUrl(this.apiBase, `projects`);
+        const httpClient = DefaultHttpClientFactory.create();
```
Comment (cdupuis) [inline lib/operations/common/GitlabRepoRef.ts]: That's not how the HttpClientFactory is supposed to be accessed as it ties it to one particular implementation. Probably better to get it from the config like `automationClientInstance().configuration.http.client.factory`.

File: lib/operations/common/GitlabRepoRef.ts [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+            body: {
+                name: `${this.repo}`,
+                visibility,
+            },
+            headers: {
+                "Private-Token": (creds as GitlabPrivateTokenCredentials).privateToken,
+                "Content-Type": "application/json",
+            },
+
+        }).then(axiosResponse => {
```
Comment (cdupuis) [inline lib/operations/common/GitlabRepoRef.ts]: axios?

File: lib/operations/common/GitlabRepoRef.ts [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+            };
+        })
+            .catch(err => {
+                logger.error(`Error attempting to raise PR. ${url} ${err}`);
+                return Promise.reject(err);
+            });
+    }
+
+    public deleteRemote(creds: ProjectOperationCredentials): Promise<ActionResult<this>> {
+        const httpClient = DefaultHttpClientFactory.create();
```
Comment (cdupuis) [inline lib/operations/common/GitlabRepoRef.ts]: Same as above.

File: lib/operations/common/GitlabRepoRef.ts [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+            });
+    }
+
+    public setUserConfig(credentials: ProjectOperationCredentials, project: Configurable): Promise<ActionResult<any>> {
+        return Promise.resolve(successOn(this));
+    }
+
+    public raisePullRequest(credentials: ProjectOperationCredentials,
+                            title: string, body: string, head: string, base: string): Promise<ActionResult<this>> {
+        const httpClient = DefaultHttpClientFactory.create();
```
Comment (cdupuis) [inline lib/operations/common/GitlabRepoRef.ts]: again

File: lib/operations/common/RepoId.ts [hunk 6]
```diff
@@ -69,7 +71,7 @@ export interface RemoteRepoRef extends RepoRef {
      */
     readonly remoteBase: string;
 
-    readonly providerType: ProviderType;
+    readonly providerType: ProviderType | string;
```
Comment (cdupuis) [inline lib/operations/common/RepoId.ts]: Why that change? We deliberately only support providers we know about.

File: lib/operations/common/gitlabRepoLoader.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+/**
+ * Materialize from gitlab
+ * @param credentials provider token
+ * @param directoryManager strategy for handling local storage
+ * @return function to materialize repos
+ */
+export function gitlabRepoLoader(credentials: ProjectOperationCredentials,
+                                 directoryManager: DirectoryManager = DefaultDirectoryManager): RepoLoader<GitProject> {
+    return repoId => {
+        // Default it if it isn't already a GitHub repo ref
```
Comment (cdupuis) [inline lib/operations/common/gitlabRepoLoader.ts]: GitHub repo ref?

File: lib/operations/common/params/GitlabTargetsParams.ts [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+        ...GitBranchRegExp,
+        required: false,
+    })
+    public sha: string = "master";
+
+    @Parameter({ description: "regex", required: false })
+    public repos: string = ".*";
+
+    get credentials(): ProjectOperationCredentials {
+        const creds: GitlabPrivateTokenCredentials = {privateToken: this.token};
```
Comment (cdupuis) [inline lib/operations/common/params/GitlabTargetsParams.ts]: formatting: space between open and closing `{ privateToken: this.token }`

File: lib/operations/common/GitlabRepoRef.ts [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+        const gitlabUrl = this.concatUrl(this.apiBase, `project/${this.owner}%2f${this.repo}`);
+        logger.debug(`Making request to '${url}' to delete repo`);
+        return httpClient.exchange(gitlabUrl, {
+            method: HttpMethod.Delete,
+            headers: {
+                "Private-Token": (creds as GitlabPrivateTokenCredentials).privateToken,
+                "Content-Type": "application/json",
+            },
+        }).then(axiosResponse => {
```
Comment (ddgenome) [inline lib/operations/common/GitlabRepoRef.ts]: Not axios

File: lib/operations/common/GitlabRepoRef.ts [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+            return {
+                target: this,
+                success: true,
+                axiosResponse,
+            };
+        })
+            .catch(err => {
```
Comment (ddgenome) [inline lib/operations/common/GitlabRepoRef.ts]: Above you put the next link of the chain on the same line, here you have a newline.

File: lib/operations/common/GitlabRepoRef.ts [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+                logger.error("Error attempting to delete repository: " + err);
+                return Promise.reject(err);
+            });
+    }
+
+    public setUserConfig(credentials: ProjectOperationCredentials, project: Configurable): Promise<ActionResult<any>> {
+        return Promise.resolve(successOn(this));
```
Comment (ddgenome) [inline lib/operations/common/GitlabRepoRef.ts]: If you are not going to implement this, perhaps it should reject?

File: lib/operations/common/GitlabRepoRef.ts [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+                title,
+                description: body,
+                source_branch: head,
+                target_branch: base,
+            },
+            headers: {
+                "Private-Token": (credentials as GitlabPrivateTokenCredentials).privateToken,
+                "Content-Type": "application/json",
+            },
+        }).then(axiosResponse => {
```
Comment (ddgenome) [inline lib/operations/common/GitlabRepoRef.ts]: Not axios

File: lib/operations/common/params/GitlabTargetsParams.ts [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+        return creds;
+    }
+
+    get description(): string {
+        return "Gitlab";
+    }
+
+    get visibility(): "public" | "private" {
+        return "public";
```
Comment (ddgenome) [inline lib/operations/common/params/GitlabTargetsParams.ts]: Does Gitlab not support private repos?

File: lib/operations/generate/GitlabRepoCreationParameters.ts [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+import { ProjectOperationCredentials } from "../common/ProjectOperationCredentials";
+import { RemoteRepoRef } from "../common/RepoId";
+import { NewRepoCreationParameters } from "./NewRepoCreationParameters";
+
+/**
+ * Parameters common to all generators that create new repositories on Gitlab
+ */
+export class GitlabRepoCreationParameters extends NewRepoCreationParameters {
+
+    @MappedParameter(MappedParameters.GitHubWebHookUrl)
```
Comment (ddgenome) [inline lib/operations/generate/GitlabRepoCreationParameters.ts]: Unfortunate names for these mapped parameters.

File: lib/operations/common/GitlabRepoRef.ts [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+        branch?: string,
+    }): GitlabRepoRef {
+        if (params.sha && !params.sha.match(GitShaRegExp.pattern)) {
+            throw new Error("You provided an invalid SHA: " + params.sha);
+        }
+        const result = new GitlabRepoRef(params.owner, params.repo, params.sha, params.rawApiBase, params.gitlabRemoteUrl, params.path);
+        result.branch = params.branch;
+        return result;
+    }
+
```
Comment (ddgenome) [inline lib/operations/common/GitlabRepoRef.ts]: GitHubRepoRef defines a `kind` property, which is nice https://github.com/atomist/automation-client/blob/87276736e56533ae235dd36fa7335906ca9ea91a/lib/operations/common/GitHubRepoRef.ts#L53
I don't think the BitBucketRepoRef does unfortunately. Maybe it should be added to RemoteRepoRef.

**Comment by cdupuis** [comment]
What is the state of this?

File: lib/operations/common/GitlabRepoRef.ts (hunk 11 above)
Comment (lievendoclo) [reply lib/operations/common/GitlabRepoRef.ts]: bitbucket implementation does the same thing, leaving this for now

**Comment by lievendoclo** [comment]
Comments processed

**Pull request merged**